### ISSUE TITLE
Factory is not thread safe

### ISSUE THREAD

**Issue by bbrk24** [Tiny-Home-Consulting/Dependiject#35]
## Checklist

- [x] I updated to the latest version and the problem persists.
- [x] I searched the [existing GitHub issues][1] and this hasn't already been reported.

## Overview

The Factory's methods are not thread-safe. Calling `register(builder:)` or `resolve(_:name:)` from multiple threads at once often results in crashes.

## Example Reproduction

```swift
await withTaskGroup(of: Void.self) { group in
    for i in 0..<100 {
        group.addTask {
            Factory.register {
                Service(.singleton, Int.self) { [i] (_) in i }
            }
        }
    }
}

await withTaskGroup(of: Void.self) { group in
    for _ in 0..<100 {
        group.addTask {
            _ = Factory.shared.resolve(Int.self)
        }
    }
}
```

## Behavior

The above code will often crash with a fatal error, EXC_BAD_ACCESS, or arithmetic error, but not always.

Either the Factory's methods should be documented as not thread-safe, or they should be enforced to always run on the same thread (e.g. `@MainActor`).

## Environment

| Software    | Version |
|:------------|:--------|
| macOS       | `12.6.0` |
| Xcode       | `14.0.1` |
| Dependiject | `0.2.0` |

[1]: https://github.com/Tiny-Home-Consulting/Dependiject/issues?q=is%3Aissue

**Comment by bbrk24** [comment]
There's a few different ways this can be solved.

The easiest way is to document it as not thread-safe. This shifts the burden of thread management from us to the caller.

Another way would be to mark the methods as `@preconcurrency @MainActor`. This pair of attributes means that ensuring thread safety falls on the caller only when called from a non-`async` function. When called in an `async` context, the function will run on the main thread and must be `await`ed. However, this is a breaking change (so it would have to go in v1.0) even if it didn't require Swift 5.6: our own example projects' tests don't compile with this change.

A third solution which meets the other two in the middle would be something like the following:
```swift
// this method is not thread-safe...
public static func register(@RegistrationBuilder builder: () -> [Registration]) {
    self.shared.registrations += builder()
}

#if swift(>=5.5)
// ...but this one is.
@available(swift 5.5)
public static func register(
    @RegistrationBuilder builder: @MainActor () -> [Registration]
) async {
    await MainActor.run {
        self.shared.registrations += builder()
    }
}
#endif
```
I'm not so worried about the duplicated code -- we can make a `private` function that both versions call.

**Comment by wboyd600** [comment]
Just noting what you said in slack in how this applies to `Factory.shared.resolve(...)` too.

**Comment by bbrk24** [comment]
I'm removing the milestone from this issue, so that the two PRs (temp fix and breaking fix) can each go on the appropriate milestone.

**Comment by wboyd600** [comment]
Sounds good!

**Comment by hmlongco** [comment]
This was discussed as an issue in [Factory](https://github.com/hmlongco/Factory) which uses recursive locks to protect the registration system and various associated caches.

Using async registration updates as mentioned above could cause issues when running tests where there's no guarantee that that registration will be completed before a service resolution is needed.

Also shifting threads to MainActor could be non-optimal if a dependency was being created on some thread other than the main thread. Structured concurrency is cool, but since there's no mechanism to ensure synchronous access implementing locks would tend to be the better solution.

**Comment by bbrk24** [comment]
@hmlongco Thank you for the feedback!

I understand the issues with `async`; that's just a band-aid for version 0.3. Internally, it synchronizes the calls using `MainActor` (as shown in the code snippet above). In version 1.0, we will have a proper fix for this that allows for thread-safe synchronous calls.

I am curious as to what exactly your thoughts are with annotating the trailing closure argument to `Service.init` as `@MainActor`. Since the intended use case is for resolving services within SwiftUI views, the most common scenario should be to call it on the main thread anyways, so it wouldn't be a performance hit in the common case. In fact, if dependency resolution needs to happen on a background thread, it would block the main thread until it completes.

Thank you again for your comment!

**Comment by bbrk24** [comment]
After discussing this with @wboyd600, we have a solution that completely works with `NSRecursiveLock`, and doesn't require threading at all. This solution can go in version 0.3.

That being said: While investigating this issue, we found an unrelated concurrency issue with `@Store`. That _will_ have to be fixed in version 1.0.
